R/counts_per_million.R

```diff
@@ -0,0 +1,49 @@
+#' Generating counts per million from SingleCellExperiment object
+#'
+#' @param input_sce_obj A SingleCellExperiment object read from RDS
+#' @param output_dir A character vector of counts per million path
+#' @param hd5_file_dir A character vector of HDF5 file path after converting from RDS
+#' @return A directory stores counts per million 
+#' @importFrom SingleCellExperiment SingleCellExperiment
+#' @importFrom dplyr tbl
+#' @importFrom glue glue
+#' @importFrom HDF5Array saveHDF5SummarizedExperiment
+#' @importFrom purrr map
+#' @noRd
+get_counts_per_million <- function(input_sce_obj, output_dir, hd5_file_dir) {
+  
+  # Create directories
+  output_dir |>  dirname() |> dir.create( showWarnings = FALSE, recursive = TRUE)
+
+  # Save SCE to the cache directory counts folder
+  input_sce_obj |> saveHDF5SummarizedExperiment(hd5_file_dir)
+  
+  data <- input_sce_obj
+  
+  # Avoid completely empty cells
+  col_sums <- colSums(as.matrix(data@assays@data$X))
+  which_to_select <- which(col_sums >0 & col_sums < Inf)
+  
+  sce <- SingleCellExperiment(list(counts_per_million = scuttle::calculateCPM(data[,which_to_select ,drop=FALSE ], assay.type = "X")))
+  rownames(sce) <- rownames(data[,which_to_select  ])
+  colnames(sce) <- colnames(data[,which_to_select  ])
+  
+  # Avoid scaling zeros
+  which_to_select_not <- setdiff(1:ncol(data), which_to_select)
```

Comment:
Rather than calculating the inverse vector, you can just use the negative of `which_to_select`. Negative indices in R return all except the specified indices.